File: 253. Meeting Rooms II/minMeetingRooms.py
from typing import List

class Solution:
  def minMeetingRooms(self, intervals: List[List[int]]) -> int:
    if not intervals:
      return 0

    start = sorted([interval[0] for interval in intervals])
    end = sorted([interval[1] for interval in intervals]) 
    
    res, count = 0, 0
    s, e = 0, 0
    
    while s < len(intervals):
      if start[s] < end[e]:
        count += 1
        s += 1
      else:
        count -= 1
        e += 1
      res = max(res, count)
      
    return res


intervals = [[0, 30], [5, 10], [15, 20]]
solution = Solution()
print("minMeetingRooms: ", solution.minMeetingRooms(intervals)) # 2

File: 253. Meeting Rooms II/minMeetingRooms.js
/**
 * @param {number[][]} intervals
 * @return {number}
 */
const minMeetingRooms = (intervals) => {
  if (!intervals.length) {
    return 0;
  }

  const starts = [];
  const ends = [];

  for (let i = 0; i < intervals.length; i++) {
    starts.push(intervals[i][0]);
    ends.push(intervals[i][1]);
  }

  starts.sort((a, b) => a - b);
  ends.sort((a, b) => a - b);

  let res = 0;
  let count = 0;
  let start = 0;
  let end = 0;

  while (start < intervals.length) {
    if (starts[start] < ends[end]) {
      count++;
      start++;
    } else {
      count--;
      end++;
    }
    res = Math.max(res, count);
  }

  return res;
};


const intervals = [[0, 30], [5, 10], [15, 20]];
console.log("minMeetingRooms: ", minMeetingRooms(intervals)); // 2
